File: backend/internal/services/shared/show_price_test.go
package shared

import (
	"testing"

	"github.com/stretchr/testify/assert"
)

// The dense-money register, pinned where every backend surface can reach it.
//
// The ICS feeds are the caller that made this urgent: a calendar entry persists
// in a subscriber's phone long after the page view that created it, so serving
// the advance half of a split price there under-reported the cost until the
// reader was at the door. The register matches the site's list surfaces exactly,
// so the number seen on /shows is the number in the calendar.
func TestShowPriceText(t *testing.T) {
	price := func(v float64) *float64 { return &v }

	tests := []struct {
		name          string
		advance, door *float64
		want          string
	}{
		{name: "no recorded price says nothing", want: ""},
		{name: "advance only renders bare", advance: price(35), want: "$35"},
		{name: "door only renders bare too", door: price(40), want: "$40"},
		{name: "a genuine pair renders both", advance: price(35), door: price(40), want: "$35/$40"},
		// Nothing stops a curator or an importer entering the same number
		// twice, and "$35/$35" spends two slots saying one thing.
		{name: "equal prices collapse", advance: price(35), door: price(35), want: "$35"},
		// Free is a price the site spells out, not an absence.
		{name: "zero is Free, not silence", advance: price(0), want: "Free"},
		{name: "a free advance with a paid door keeps both", advance: price(0), door: price(10), want: "Free/$10"},
		// Whole amounts drop the cents; a real fractional price keeps them.
		// "$%.0f" alone printed 12.50 as "$12" and, worse, printed a fifty-cent
		// door as "$0" — free, in a subscriber's calendar. Mirrors formatPrice
		// in frontend/lib/utils/formatters.ts.
		{name: "whole amounts drop the cents", advance: price(35), want: "$35"},
		{name: "a fractional price keeps them", advance: price(12.5), want: "$12.50"},
		{name: "and is never rounded", advance: price(12.75), want: "$12.75"},
		{name: "a sub-dollar price is not free", advance: price(0.5), want: "$0.50"},
		// Unusual but not impossible, and the order is stated rather than
		// sorted: advance first, whichever number is larger.
		{name: "a cheaper door keeps advance first", advance: price(40), door: price(30), want: "$40/$30"},
	}

	for _, tc := range tests {
		t.Run(tc.name, func(t *testing.T) {
			assert.Equal(t, tc.want, ShowPriceText(tc.advance, tc.door))
		})
	}
}

File: frontend/lib/utils/showPrice.test.ts
import { describe, it, expect } from 'vitest'
import { statedShowPrices, showPriceLabel } from './showPrice'

describe('statedShowPrices', () => {
  it('says nothing when neither price is recorded', () => {
    expect(statedShowPrices({ price: null, door_price: null })).toEqual([])
    expect(statedShowPrices({})).toEqual([])
  })

  it('returns the single price a show records, advance or door', () => {
    expect(statedShowPrices({ price: 35, door_price: null })).toEqual([35])
    expect(statedShowPrices({ price: null, door_price: 40 })).toEqual([40])
  })

  it('returns a genuine pair advance-first', () => {
    expect(statedShowPrices({ price: 35, door_price: 40 })).toEqual([35, 40])
  })

  // Nothing constrains a curator or an importer from entering the same number
  // twice, and two slots saying one thing reads as a rendering bug.
  it('collapses equal prices to one', () => {
    expect(statedShowPrices({ price: 35, door_price: 35 })).toEqual([35])
  })

  // Zero is a price the site spells "Free", so it must survive as a value
  // rather than falling to the empty case with a truthiness guard.
  it('treats zero as a price, not as silence', () => {
    expect(statedShowPrices({ price: 0, door_price: null })).toEqual([0])
    expect(statedShowPrices({ price: 0, door_price: 10 })).toEqual([0, 10])
  })

  // A door price cheaper than the advance price is unusual but not impossible,
  // and the order stays advance-first so the pair never has to be re-sorted to
  // be read.
  it('keeps advance first even when the door is cheaper', () => {
    expect(statedShowPrices({ price: 40, door_price: 30 })).toEqual([40, 30])
  })
})

describe('showPriceLabel', () => {
  it('is null when no price is recorded', () => {
    expect(showPriceLabel({ price: null, door_price: null })).toBeNull()
  })

  it('renders a lone price bare, with no qualifier to disambiguate from', () => {
    expect(showPriceLabel({ price: 35, door_price: null })).toEqual({ text: '$35' })
    expect(showPriceLabel({ price: null, door_price: 40 })).toEqual({ text: '$40' })
    expect(showPriceLabel({ price: 0, door_price: null })).toEqual({ text: 'Free' })
  })

  it('renders a pair in the dense-list register with a spelled-out title', () => {
    expect(showPriceLabel({ price: 35, door_price: 40 })).toEqual({
      text: '$35/$40',
      title: '$35 advance, $40 at the door',
    })
  })

  // The title is what a screen reader announces instead of "35 slash 40", so
  // its absence on the single-price case is deliberate rather than an omission:
  // there is nothing to disambiguate.
  it('carries no title when there is only one number', () => {
    expect(showPriceLabel({ price: 35, door_price: 35 })?.title).toBeUndefined()
  })

  it('spells cents when a price really has them', () => {
    expect(showPriceLabel({ price: 12.5, door_price: null })).toEqual({ text: '$12.50' })
  })
})
